cpp/src/stream_compaction/distinct.cu

```diff
@@ -59,12 +60,24 @@ rmm::device_uvector<size_type> get_distinct_indices(table_view const& input,
   auto const row_hasher = cudf::experimental::row::hash::row_hasher(preprocessed_input);
   auto const key_hasher = experimental::compaction_hash(row_hasher.device_hasher(has_nulls));
 
-  auto const row_comp  = cudf::experimental::row::equality::self_comparator(preprocessed_input);
-  auto const key_equal = row_comp.equal_to(has_nulls, nulls_equal);
+  auto const row_comp = cudf::experimental::row::equality::self_comparator(preprocessed_input);
 
   auto const pair_iter = cudf::detail::make_counting_transform_iterator(
     size_type{0}, [] __device__(size_type const i) { return cuco::make_pair(i, i); });
-  map.insert(pair_iter, pair_iter + input.num_rows(), key_hasher, key_equal, stream.value());
+
+  auto const insert_keys = [&](auto const value_comp) {
+    auto const key_equal = row_comp.equal_to(has_nulls, nulls_equal, value_comp);
+    map.insert(pair_iter, pair_iter + input.num_rows(), key_hasher, key_equal, stream.value());
+  };
+
+  if (nans_equal == nan_equality::ALL_EQUAL) {
+    using nan_equal_comparator =
+      cudf::experimental::row::equality::nan_equal_physical_equality_comparator;
+    insert_keys(nan_equal_comparator{});
+  } else {
+    using nan_unequal_comparator = cudf::experimental::row::equality::physical_equality_comparator;
+    insert_keys(nan_unequal_comparator{});
+  }
```

Comment:
```suggestion
  auto const nan_comparator = [&](){
    if (nans_equal == nan_equality::ALL_EQUAL) {
      return cudf::experimental::row::equality::nan_equal_physical_equality_comparator{};
    }
    return cudf::experimental::row::equality::physical_equality_comparator{};
  }();
  auto const key_equal = row_comp.equal_to(has_nulls, nulls_equal, nan_comparator);
  
  map.insert(pair_iter, pair_iter + input.num_rows(), key_hasher, key_equal, stream.value());
```

This is shorter and easier to follow IMO since we don't need to jump back and forth looking for the predefined lambda function. Also, defining an alias used only once seems redundant.

Reply:
Any function must return object of the same type, with the exception that different return types must be enclosed in compile-time mutually exclusive code blocks (like `if constexpr`).

Reply:
Runtime information indeed!

Thank you both for the clarification. Resolving this.